iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_expect.cpp

```diff
@@ -72,15 +72,12 @@ TYPED_TEST(FunctionalInterface_test, ExpectDoesNotCallTerminateWhenObjectIsValid
 template <typename FactoryType, typename SutType, typename ExpectCall>
 void ExpectDoesCallTerminateWhenObjectIsInvalid(const ExpectCall& callExpect)
 {
-    bool wasErrorHandlerCalled = true;
     SutType sut = FactoryType::createInvalidObject();
     {
-        auto handle = iox::ErrorHandlerMock::setTemporaryErrorHandler<iox::HoofsError>(
-            [&](auto, auto) { wasErrorHandlerCalled = true; });
-        callExpect(sut);
+        auto handle =
+            iox::ErrorHandlerMock::setTemporaryErrorHandler<iox::HoofsError>([&](auto, auto) { std::terminate(); });
+        EXPECT_DEATH(callExpect(sut), ".*");
```

Comment:
Sadly no. The reason is undefined behavior in a functional interface construct. The contract is either return a reference to the underlying value or die. And this has to be testable.

Assume you have an `optional` or an `expected` which does not contain a value.
Now we test `expect` for those objects. When they do not contain an error the expect has to terminate otherwise we have to dereference a nullptr or some garbage memory since they do not contain any value. So the only option is to terminate and check this.

Another alternative would of course be an exception in the custom user callback. This would make such code much more testable and would solve a lot of issues. But we have to remove `noexcept` before that in all methods and functions.